aexpect/session_ops.py

```diff
@@ -0,0 +1,598 @@
+                           .format(tarball, target_dir, output))
+
+
+def ls(session, dir_name):  # pylint: disable=C0103
+    """
+    Run `ls` in given directory through a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str dir_name: name of directory
+    :returns: names of files in dir (without path components)
+    :rtype: [str]
+
+    Just like :py:func:`os.listdir`, does not include file names starting with
+    dot (`'.'`)
+    """
+    cmd = 'ls -1UNq {}'.format(quote(dir_name))
+    status, output = session.cmd_status_output(cmd)
+    if output:
+        output = output.strip()
+    if status == 2:      # probably just nothing found
+        return []
+    if status != 0:
+        raise RuntimeError('Failed to ls {}: {}'
+                           .format(dir_name, output))
+    return output.splitlines()
+
+
+def glob(session, glob_pattern):
+    """
+    Find files matching given pattern through a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str glob_pattern: pattern of filenames where `*`, `?` and such match
+                             any string or not-empty-string and so on.
+    :returns: file names matching pattern, including path (like glob.glob)
+    :rtype: [str]
+    :raises: :py:class:`RuntimeError` if tar ls-command returned non-null and
+             non-2 status
+
+    Much like :py:func:`glob.glob` but implemented using `ls`.
+
+    The glob matching is done locally, not with `ls` since quoting also makes
+    the glob characters '*' and '?' regular characters, so are not interpreted
+    any more in shell.
+    """
+    path_part, name_pattern = path_split(glob_pattern)
+    all_filenames = ls(session, path_part)
+
+    # convert pattern: escape \ . ? *
+    pattern = re.escape(name_pattern)
+
+    # convert pattern: convert shell globs to regex
+    pattern = pattern.replace(r'\*', '.*').replace(r'\?', '.')
+
+    # convert pattern: require match whole file name; compile for speed-up
+    pattern = re.compile(pattern + '$')
+    return [join(path_part, filename) for filename in all_filenames
+            if pattern.match(filename)]
+
+
+def cat(session, filename):
+    """
+    Get contents of a text file from a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str filename: full path of file
+    :returns: file contents
+    :rtype: str
+    :raises: :py:class:`RuntimeError` if cat command fails
+
+    Should only be used for very short files without tabs or other fancy
+    contents. Otherwise better download file or use some other method.
+    """
+    cmd = 'cat ' + quote(filename)
+    status, output = session.cmd_status_output(cmd)
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError('Failed to cat {}: {}'.format(filename, output))
+    return output
+
+
+def tempdir(session):
```

Comment:
s/tempdir/create_tempdir/

Reply:
In case of ls you are calling the function `ls()`. But here you are using `tempdir` instead of `mktemp`, any reason?

Reply:
Probably not, will correct this and write back.

Reply:
Originally some of our devs started with the idea of mimicking the remote command as a function name (hence `ls`) but later on I realized that moving away from the one-to-one name and into a more descriptive function name is better for the API. So I think your `create_tempdir` suggestion is good.

Reply:
I went for something in the middle like `make_tempdir`